File: src/nric.c
#include <stdio.h>
#include <stdlib.h> /* atoi */
#include <string.h> /* strlen */

int multiples[7] = { 2, 7, 6, 5, 4, 3, 2 };

int is_nric_valid(char* the_nric) {
	char* first;
	char* last;

	int numeric;
	int i;
	char outputs[11] = { 'J', 'Z', 'I', 'H', 'G', 'F', 'E', 'D', 'C', 'B', 'A' };
	char outputs2[11] = { 'G', 'F', 'E', 'D', 'C', 'B', 'A', 'J', 'Z', 'I', 'H' };
	char *output;

	if (the_nric == NULL) {
		return 0;
	}

	if (strlen(the_nric) != 9) {
		return 0;
	}

	first = &the_nric[0];
	last = &the_nric[strlen(the_nric) - 1];

	if (*first != 'S' && *first != 'T') {
		return 0;
	}

	numeric = atoi(&the_nric[1]);

	if (*first == 'S') {
		output = outputs;
	} else {
		output = outputs2;
	}

	return check_mod_11(last, numeric, output);

}

int is_fin_valid(char* the_fin) {
	char* first;
	char* last;

	int numeric;
	int i;
	char outputs[11] = {  'X', 'W', 'U', 'T', 'R', 'Q', 'P', 'N', 'M', 'L', 'K' };
	char outputs2[11] = { 'R', 'Q', 'P', 'N', 'M', 'L', 'K', 'X', 'W', 'U', 'T' };
	char* output;

	if (the_fin == NULL) {
		return 0;
	}

	if (strlen(the_fin) != 9) {
		return 0;
	}

	first = &the_fin[0];
	last = &the_fin[strlen(the_fin) - 1];

	if (*first != 'F' && *first != 'G') {
		return 0;
	}

	numeric = atoi(&the_fin[1]);

	if (*first == 'F') {
		output = outputs;
	} else {
		output = outputs2;
	}

	return check_mod_11(last, numeric, output);

}

int check_mod_11 (char* last, int numeric, char* outputs) {
	int total = 0;
	int count = 0;
	int len = 7;

	while (numeric != 0) {
		total += (numeric % 10) * multiples[len - (1 + count)];
		count ++;

		numeric /= 10;
	}

	if (*last == outputs[total % 11]) {
		return 1;
	}

	return 0;
}

main() {

	int a = is_nric_valid("G6046409Q");

	printf("0 == %d\n", a);

	a = is_nric_valid("S8944027J");

	printf("1 == %d\n", a);

	a = is_fin_valid("G6046409Q");

	printf("1 == %d\n", a);

	a = is_fin_valid("G604a409Q");

	printf("0 == %d\n", a);

	a = is_nric_valid("G604a409Q");

	printf("0 == %d\n", a);
}

File: src/nric.scala
class NricValidator () {

  val multiples = 2 :: 7 :: 6 :: 5 :: 4 :: 3 :: 2 :: Nil

  def isNricValid(theNric: String): Boolean = {

    if (theNric == Nil || theNric.length != 9)
      return false

    val first = theNric(0)
    val last = theNric(theNric.length - 1)
    var outputs: List[String] = Nil

    if (!(Array('S', 'T') contains first))
      return false

    var numeric: Int = 0
    try {
      numeric= theNric.slice(1, theNric.length - 1).toInt
    } catch { 
      case e => return false 
    }

    checkMod11(last, numeric, getArray(first) )
  }

  def isFinValid(theFin: String): Boolean = {
    if (theFin == Nil || theFin.length != 9)
      return false

    val first = theFin(0)
    val last = theFin(theFin.length -1)
    
    if (!(Array('F', 'G') contains first))
      return false
    
    var numeric: Int = 0
    try {
      numeric = theFin.slice(1, theFin.length - 1).toInt
    } catch { 
      case e =>return false 
    }
    
    checkMod11(last, numeric, getArray(first))
  }

  def getArray(first: Char): List[Char] = { 
    if (first == 'S')
      'J' :: 'Z' :: 'I' :: 'H' :: 'G' :: 'F' :: 'E' :: 'D' :: 'C' :: 'B' :: 'A' :: Nil
    else if (first == 'F')
      'X' :: 'W' :: 'U' :: 'T' :: 'R' :: 'Q' :: 'P' :: 'N' :: 'M' :: 'L' :: 'K' :: Nil
    else if (first == 'G')
      'R' :: 'Q' :: 'P' :: 'N' :: 'M' :: 'L' :: 'K' :: 'X' :: 'W' :: 'U' :: 'T' :: Nil
    else
      'G' :: 'F' :: 'E' :: 'D' :: 'C' :: 'B' :: 'A' :: 'J' :: 'Z' :: 'I' :: 'H' :: Nil
  }

  def checkMod11(last: Char, numeric: Int, outputs: List[Char]): Boolean = {
    var total = 0
    var count = 0
    var num = numeric

    while (num != 0) {
      total += (num % 10) * multiples(multiples.length - (1 + count))
      count += 1
      num /= 10
    }

    last == outputs(total % 11)

  }
}


object NricValidatorTest {
  def main(args: Array[String]) {

    val validator = new NricValidator()
    
    assert ( validator.isNricValid("S8944027J") == true) // True
    assert ( validator.isNricValid("G6046409Q") == false) // False
    assert ( validator.isNricValid("123") == false)       // False
    assert ( validator.isFinValid("G6046409Q") == true)  // True
    assert ( validator.isFinValid("123") == false)        // False
    assert ( validator.isFinValid("G6046a09Q") == false)  // False

  }
}
